hasktorch/src/Torch/NN.hs

```diff
@@ -117,40 +118,57 @@ class Parameterized f where
   default _replaceParameters :: (Generic f, GParameterized (Rep f)) => f -> ParamStream f
   _replaceParameters f = to <$> _gReplaceParameters (from f)
 
+  replaceDevice :: Device -> f -> f
+  default replaceDevice :: (Generic f, GParameterized (Rep f)) => Device -> f -> f
+  replaceDevice dev f = to $ gReplaceDevice dev (from f)
+
+defaultReplaceDevice :: Parameterized a => Device -> a -> a
+defaultReplaceDevice dev f = replaceParameters f $ map (IndependentTensor . (_toDevice dev) . toDependent) $ flattenParameters f
+
 replaceParameters :: Parameterized f => f -> [Parameter] -> f
 replaceParameters f params =
   let (f', remaining) = runState (_replaceParameters f) params
    in if null remaining
         then f'
         else error "Some parameters in a call to replaceParameters haven't been consumed!"
 
+instance Parameterized a => ToDevice a where
+  toDevice = replaceDevice
+
 instance Parameterized Tensor where
   flattenParameters _ = []
   _replaceParameters = return
+  replaceDevice = _toDevice
```

Comment:
replaceDevice is similar to _replaceParameters.
Tensor type is not a parameter, but it is the target of the move.
Parameter type is a parameter and the target of the move.